libs/payments/iap/src/lib/apple/apple-iap.service.ts

```diff
@@ -0,0 +1,168 @@
+            originalTransactionId,
+          }
+        );
+        throw libraryError;
+      }
+    }
+    // Unlike Google Play subscriptions, We don't need to check if the purchase is
+    // registerable (i.e. not terminal), since the original transaction ID is
+    // the primary key in Firestore, and it can be "revived at any time if the customer
+    // returns to any SKU in that subscription group".
+    // See https://developer.apple.com/forums/thread/704167.
+
+    // STEP 2. Check if the purchase has been registered to a user. If it is, then return conflict error to our caller.
+    if (purchase.userId === userId) {
+      // Purchase record already registered to the target user. We'll do nothing.
+      return purchase;
+    } else if (purchase.userId) {
+      // this.log.info('Purchase already registered', { purchase });
+      // Purchase record already registered to different user. Return 'conflict' to caller
+      const libraryError = new AppleIapConflictError(
+        'Purchase has been registered to another user'
+      );
+      this.log.error('registerToUserAccount.PurchaseUpdateError.Conflict', {
+        err: libraryError,
+        originalTransactionId,
+      });
+      throw libraryError;
+    }
+
+    // STEP 3: Register purchase to the user
+    await this.appleIapPurchaseManager.forceRegisterToUser(
+      originalTransactionId,
+      userId
+    );
+
+    return purchase;
+  }
+
+  /**
+   * Update Firestore with the latest subscription status information from
+   * the App Store Server.
+   *
+   * See https://developer.apple.com/documentation/appstoreservernotifications/notificationtype
+   * and https://developer.apple.com/documentation/appstoreservernotifications/subtype
+   */
+  async processNotification(
+    bundleId: string,
+    originalTransactionId: string,
+    notification: DecodedNotificationPayload
+  ): Promise<AppStoreSubscriptionPurchase | null> {
+    // Type-guard to ensure the notification is for a subscription
+    if (
+      NOTIFICATION_TYPES_FOR_NON_SUBSCRIPTION_PURCHASES.includes(
+        notification.notificationType
+      )
+    ) {
+      return null;
+    }
+
+    // We can safely ignore Subscribed - InitialBuy notifications, because
+    // with a new subscription, our iOS app will send the same original
+    // transaction ID to the server for verification.
+    if (
+      notification.notificationType === NotificationType.Subscribed &&
+      notification?.subtype === NotificationSubtype.InitialBuy
+    ) {
+      return null;
+    }
+
+    // For other types of notifications, we query the App Store Server API
+    // to update our purchase record cache in Firestore
+    return this.querySubscriptionPurchase(
+      bundleId,
+      originalTransactionId,
+      notification.notificationType,
+      notification?.subtype
+    );
+  }
+
+  /**
+   * Retrieves a list of known subscription purchases for the customer.
+   */
+  async getSubscriptions(
```

Comment:
Do you know where we'll be consuming this? I believe this can be eliminated, since this would be handled by some other service.